torchmultimodal/utils/generate.py

```diff
@@ -0,0 +1,212 @@
+        # Attention mask is required to avoid attending to future positions along the sequence.
+        # See :class:`~torchmultimodal.models.gpt.RightShift` for more implementation details.
+        self.model.fwd(  # type: ignore
+            in_tokens=in_tokens,
+            attn_mask=attn_mask,
+            use_cache=use_cache,
+            causal=causal,
+            right_shift=True,
+        )
+
+        # Take the last token of the input modality as the "sos" token for the ouput modality
+        samples = [in_tokens[:, -1:]]
+        idx = 0
+        # Attention mask is not required as the cached key/value sequence is only up to the
+        # current `decode_step`
+        while idx < self.max_seq_len:
+            # decode_step is w.r.t. the total sequence concat from in and out
+            # Shift by in_seq_len because we start from the out sequence
+            decode_step = idx + in_seq_len
+            if idx == 0:
+                out = self.model(
+                    in_tokens=samples[-1],
+                    in_pos_ids=torch.tensor([decode_step - 1]).unsqueeze(0),
+                    logits_mask=logits_mask[[decode_step]],
+                    use_cache=use_cache,
+                    causal=causal,
+                    right_shift=False,
+                )
+            else:
+                out = self.model(
+                    out_tokens=samples[-1],
+                    out_pos_ids=torch.tensor([idx - 1]).unsqueeze(0),
+                    logits_mask=logits_mask[[decode_step]],
+                    use_cache=use_cache,
+                    causal=causal,
+                    right_shift=False,
+                )
+
+            assert hasattr(out, "logits"), f"{type(out)} does not have field 'logits'"
+            logits = out.logits
+
+            logits_view = logits.view(-1, logits.shape[-1])  # (b, num_tokens)
+            if top_k is not None or top_p is not None:
+                logits_view = top_k_top_p_filtering(
+                    logits_view, top_k=top_k, top_p=top_p
+                )
+            probs = F.softmax(logits_view, dim=-1)
+            samples.append(torch.multinomial(probs, 1))  # (b, 1)
+            idx += 1
+
+        samples = torch.cat(
+            samples[1:], dim=1
+        )  # remove the "sos" token: (b, out_seq_len)
+        samples = self.model.decode(samples)  # type: ignore
+
+        return samples
+
+
+def get_logits_mask(
+    in_seq_len: int, out_seq_len: int, num_in_tokens: int, num_out_tokens: int
+) -> Tensor:
+    """
+    Generates masks to restrict prediction to be made from tokens of the same modality.
+    """
+    total_seq_len = in_seq_len + out_seq_len
+    num_tokens = num_in_tokens + num_out_tokens
+    logits_mask = torch.ones(total_seq_len, num_tokens)
+    logits_mask[in_seq_len:, :num_in_tokens] = 0
+    logits_mask[:in_seq_len, num_in_tokens:] = 0
+
+    return logits_mask  # (total_seq_len, total_num_tokens)
+
+
+def top_k_top_p_filtering(
```

Comment:
This utility is a bit confusing to me... so users will want to perform both top-k and top-p filtering for the same set of predictions? Aren't these two non-commutative operations?

Reply:
I think the more common scenarios would be either one of the two. But there's nothing wrong to pick top-k logits before picking amongst the remainder the top-p cumulative probability. what do you think?

Reply:
Yeah I think it's reasonable to use both. My comment is more around how we structure it. Since the order in which we apply top-k and top-p matters, maybe it's better to not combine the two into a single function like this, and instead just apply them separately in the main script.

Reply:
>Since the order in which we apply top-k and top-p matters

Thought I missed the point of this question: looks like it's a convention to apply top-k before top-p if both knobs are not `None`. I am not sure whether it's just out of convenience than strong mathematical reasons. My guess is that if we do `top_p` first we won't know for sure how many tokens will be left (as the filter reduce the samples by cumulative probability) for us to prescribe a meaningful `top_k`.